### ISSUE TITLE
MatrixFree::loop with hp not working

### ISSUE THREAD

**Issue by Rombur** [dealii/dealii#12102]
I am successfully using `MatrixFree::cell_loop()` with `FE_Nothing` and I wanted to use `MatrixFree::loop()` to impose Neumann boundary condition but that does not work. Is it something that we support or not? As an example of the problems I've encountered, I've modified `matrix_free/matrix_vector_hp.cc` using the following patch:
```
diff --git a/tests/matrix_free/matrix_vector_hp.cc b/tests/matrix_free/matrix_vector_hp.cc
index ac18292b62..3cbbe05a77 100644
--- a/tests/matrix_free/matrix_vector_hp.cc
+++ b/tests/matrix_free/matrix_vector_hp.cc
@@ -91,11 +91,34 @@ public:
                                                     subrange_deg);
   }
 
+  void
+  face_apply(const MatrixFree<dim, Number> &,
+             Vector<Number> &,
+             const Vector<Number> &,
+             const std::pair<unsigned int, unsigned int> &) const
+  {
+    // no-op
+  }
+
+  void
+  boundary_apply(const MatrixFree<dim, Number> &,
+                 Vector<Number> &,
+                 const Vector<Number> &,
+                 const std::pair<unsigned int, unsigned int> &) const
+  {
+    // no-op
+  }
+
   void
   vmult(Vector<Number> &dst, const Vector<Number> &src) const
   {
     dst = 0;
-    data.cell_loop(&MatrixFreeTestHP<dim, Number>::local_apply, this, dst, src);
+    data.loop(&MatrixFreeTestHP<dim, Number>::local_apply,
+              &MatrixFreeTestHP<dim, Number>::face_apply,
+              &MatrixFreeTestHP<dim, Number>::boundary_apply,
+              this,
+              dst,
+              src);
   };
 
 private:
@@ -194,6 +217,10 @@ test()
   MatrixFree<dim, number>                          mf_data;
   typename MatrixFree<dim, number>::AdditionalData data;
   data.tasks_parallel_scheme = MatrixFree<dim, number>::AdditionalData::none;
+  data.mapping_update_flags_inner_faces =
+    update_JxW_values | update_quadrature_points;
+  data.mapping_update_flags_boundary_faces =
+    update_JxW_values | update_quadrature_points;
   mf_data.reinit(dof, constraints, quadrature_collection_mf, data);
   MatrixFreeTestHP<dim, number> mf(mf_data);
```
I get the following error:
```
An error occurred in line <5976> of file <../include/deal.II/fe/fe_values.h> in function                                                                                                                                                     
    const DerivativeForm<1, dim, spacedim> &dealii::FEValuesBase<2, 2>::jacobian(const unsigned int) const [dim = 2, spacedim = 2]                                                                                                           
The violated condition was:                                                                                                                                                                                                                  
    static_cast<typename ::dealii::internal::argument_type<void( typename std::common_type<decltype(i), decltype(this->mapping_output.jacobians.size())>::type)>:: type>(i) < static_cast<typename ::dealii::internal::argument_type<void( ty
pename std::common_type<decltype(i), decltype(this->mapping_output.jacobians.size())>::type)>:: type>(this->mapping_output.jacobians.size())                                                                                                 
Additional information:                                                                                                                                                                                                                      
    Index 4 is not in the half-open range [0,4).  
```
In my own code, I use `LA::distributed::Vector` instead of serial `Vector`. In that case, I get an error earlier from `dealii::internal::MatrixFreeFunctions::DoFInfo::compute_tight_partitioners`.
Should I expect code that works with `cell_loop` to also work with `loop` or not?

**Comment by peterrum** [comment]
As far as I remember, we do not support MatrixFree+hp+DG(aka. MatrixFree::loop). But let me check tomorrow what is missing for your use case (I have recently misused and extended the infrastructure so that we can do MatrixFree+DG methods also for mixed meshes).

For sake of simplicity, could you post the complete test program. Thx!

**Comment by Rombur** [comment]
[Here](https://github.com/Rombur/dealii/commit/ab41cb6813f6a06e10cd6efe2897063c23c02df5#diff-609b295dde73e1cdce24e0b668eba5394abf77065099769a04b981e70706a74a) is the commit containing the file. Like I said if you use distributed Vector then the code crashes even earlier, even one processor. It's not a big for me if don't support it but it would be great if we had an assert in the code.

**Comment by kronbichler** [comment]
It's right, we don't support it right now, but I think the combination `FE_Q`/`FE_DGQ` plus `FE_Nothing` should be relatively simple to address, because we merely need to consider the fe nothing as empty cells with some magic flags for the face detection. Whereas true hp functionality is considerably more work because the interfaces need to adjust their degrees for cross-element vectorization, quadrature formulas need to be synched, etc.

**Comment by Rombur** [comment]
I see that makes sense. Personally, I only care about `FE_Nothing`. I'd be happy to work on it but I will need some guidance.

**Comment by peterrum** [comment]
> It's right, we don't support it right now, but I think the combination FE_Q/FE_DGQ plus FE_Nothing should be relatively simple to address, because we merely need to consider the fe nothing as empty cells with some magic flags for the face detection. Whereas true hp functionality is considerably more work because the interfaces need to adjust their degrees for cross-element vectorization, quadrature formulas need to be synched, etc.

Yes. This is what I also think: in the case of mixed meshes we know that both sides have the same number of quadrature points, here we can also simply decide which quadrature rule to use. This is what I was referring to as "misusing".

> I see that makes sense. Personally, I only care about FE_Nothing. I'd be happy to work on it but I will need some guidance.

Let me check where you are thrown out. I'll get to you back later!

**Comment by peterrum** [comment]
I am kicked out here (during setup of the mapping data in the quadrature points):  
https://github.com/dealii/dealii/blob/42e48fa22db833282dd8a38858c44dda94fd03dc/include/deal.II/matrix_free/mapping_info.templates.h#L2084-L2087

which makes sense since `n_q_points` is selected for the interior face (which happens to have non FE_Nothing), but we here we  have an exterior face (which happens to have a FE_Nothing assigned to).

Personally, I would like to keep the code here as it is, but throw out any faces and face-face pairs that have on one side a `FE_Nothing` assigned to, i.e. ignore them in `internal::MatrixFreeFunctions::FaceInfo`. An issue I see here is that `internal::MatrixFreeFunctions::FaceInfo::cell_and_face_to_plain_faces` might be a bit confused (since some faces are missing or they are interpreted as boundary faces - but is there anyone who wants to do cell-centric loops in this context?). Anyhow, this approach would mean that those faces are not visited by the matrix-free loop. For me this sounds reasonable.

**Comment by peterrum** [comment]
A face is created here: 

https://github.com/dealii/dealii/blob/42e48fa22db833282dd8a38858c44dda94fd03dc/include/deal.II/matrix_free/face_setup_internal.h#L947-L1021

and they are inserted into the global data structures here:

https://github.com/dealii/dealii/blob/42e48fa22db833282dd8a38858c44dda94fd03dc/include/deal.II/matrix_free/face_setup_internal.h#L770-L935

I am happy to help out. However, we should agree if the proposed solution is acceptable or a bit too hacky...

**Comment by Rombur** [comment]
> Personally, I would like to keep the code here as it is, but throw out any faces and face-face pairs that have on one side a FE_Nothing assigned to, i.e. ignore them in internal::MatrixFreeFunctions::FaceInfo.

Yes, I think that makes sense. Another way would be to allow MatrixFree to take a predicate similar to what I did for CUDA https://github.com/dealii/dealii/pull/11780 This allows me to only loop over cells of a given fe index and thus, I can skip all the cells with FE_Nothing. For CUDA it was pretty easy to do since we create a graph that contains all the cells before we start looping and so the predicate determines which cells are added to the graph.

**Comment by peterrum** [comment]
>  Another way would be to allow MatrixFree to take a predicate similar to what I did for CUDA #11780 This allows me to only loop over cells of a given fe index and thus, I can skip all the cells with FE_Nothing. For CUDA it was pretty easy to do since we create a graph that contains all the cells before we start looping and so the predicate determines which cells are added to the graph.

This is a bit more complicated here, since we also have to take care of faces. But let's wait for @kronbichler how he would tackle this issue and/or he agrees.

**Comment by gfcas** [comment]
May I ask: I consider multi-physical problems, where the solution has a jump over the interface. Therefore I use the comibantion FE_QxFE_Nothing and FE_NothingxFE_Q (according to step-46). At the interface the coupling is described by an nonlinear Neumann condition. So this would be a missing part for the matrix-free implementation of my problem right? In that case I would also be interested in fixing that issue.

**Comment by Rombur** [comment]
Yes, you will hit the same problem. We need to decide how we should fix this. @kronbichler any preference? Should we ignore FE_Nothing faces in `internal::MatrixFreeFunctions::FaceInfo` or should we loop only on cells that do not use FE_Nothing.

**Comment by bergbauer** [comment]
I am also interested in this feature and thought about tackling it in the near future.

**Comment by sebproell** [comment]
I am also interested in this feature. I want to evaluate a boundary condition on faces that separate `FE_Nothing` from (in my case) `FE_Q`s. I am not sure if I understood the discussion correctly, but for me faces that border `FE_Nothing` would need to be inlcuded in the loop.

**Comment by kronbichler** [comment]
Let us remove this from the current release milestone. It would be great to soon provide a fix, but I fear we won't manage to do it in the next few days.

**Comment by Rombur** [comment]
In #14014, we said that the problem I reported here is fixed but we have only partially fix it. The problem appears if you use a different quadrature for `FE_Nothing` than for the `FE_Q`. For some faces you hit a bug in the code below 

https://github.com/dealii/dealii/blob/1c182b617df9171bee39aae0ef9c456713a6332a/include/deal.II/matrix_free/mapping_info.templates.h#L1880-L1883

because `n_q_points` is larger than the number of jabobians in this cell. The test in #14014 doesn't catch the bug because it only uses a single quadrature. In my code I only see the problem when using multiple processors but I am not sure that the bug only appears in the distributed case. Anyway, I don't know if there is an easy to fix this but it would good if we could have an assert somewhere.

**Comment by kronbichler** [comment]
I believe we can add a simple check regarding the quadrature formula (using a single quadrature formula, as suggested by #14014) - @peterrum, can you confirm, then we simply add the assertion, and do the rest of this issue after the release.

**Comment by Rombur** [comment]
No, I can try to write one tomorrow

**Comment by bangerth** [comment]
@Rombur Ping?

Can I propose that we postpone this issue to the next release?

**Comment by kronbichler** [comment]
Let us postpone this issue; while we will address it, it has been open long enough that it will not make a difference. After all, the short-term fix would have been a better assertion.
